drivers: wifi: nrf_wifi: Add P2P discovery and Client support

rado17 opened the pull request:
Add P2P support and bug fixes. New Wi-Fi management API and a shell APIs are added.

rado17 committed:
drivers: wifi: nrf_wifi: Set SSID for P2P discovery

Use `ssids` instead of `filter_ssids` to set the SSID in probe
requests. `filter_ssids` are to filter scan results to include
only the specified SSIDs.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

krish2718 commented on drivers/wifi/nrf_wifi/src/wpa_supp_if.c:
```diff
@@ -1798,6 +1798,9 @@ int nrf_wifi_supp_get_capa(void *if_priv, struct wpa_driver_capa *capa)
 	if (IS_ENABLED(CONFIG_NRF70_AP_MODE)) {
 		capa->flags |= WPA_DRIVER_FLAGS_AP;
 	}
+	if (IS_ENABLED(CONFIG_NRF70_P2P_MODE)) {
+		capa->flags |= WPA_DRIVER_FLAGS_P2P_CAPABLE;
+	}
```
move to a separate commit.

krish2718 commented on drivers/wifi/nrf_wifi/src/wpa_supp_if.c:
```diff
@@ -545,16 +545,16 @@ int nrf_wifi_wpa_supp_scan2(void *if_priv, struct wpa_driver_scan_params *params
+		scan_info->scan_params.num_scan_ssids = params->num_ssids;
 
-		for (indx = 0; indx < params->num_filter_ssids; indx++) {
+		for (indx = 0; indx < params->num_ssids; indx++) {
 			memcpy(scan_info->scan_params.scan_ssids[indx].nrf_wifi_ssid,
-			       params->filter_ssids[indx].ssid,
-			       params->filter_ssids[indx].ssid_len);
+			       params->ssids[indx].ssid,
+			       params->ssids[indx].ssid_len);
 
 			scan_info->scan_params.scan_ssids[indx].nrf_wifi_ssid_len =
-				params->filter_ssids[indx].ssid_len;
+				params->ssids[indx].ssid_len;
 		}
 	}
```
update commit to reflect that it's a bug as filter_ssids are meant for filtering scan results. (unrelated to P2P)

kapbh committed:
net: wifi: Add Wi-Fi direct P2P discovery API support

Add supplicant api and  mgmt ops support for P2P discovery.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

rado17 committed:
drivers: wifi: nrf_wifi: Set P2P capability

In P2P mode, inform supplicant that the driver is P2P capable.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

krish2718 reviewed (commented):
Please separate API and Shell in to two commits

krish2718 commented on samples/net/wifi/shell/overlay-p2p.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_WIFI_NM_WPA_SUPPLICANT_P2P=y
+CONFIG_WPA_CLI=y
+CONFIG_NRF70_P2P_MODE=y
```
this doesn't belong here and should be auto-enabled by CONFIG_WIFI_NM_WPA_SUPPLICANT_P2P

krish2718 reviewed (changes requested):
Also, this needs a doc update in Wi-Fi API with a separate section.

kapbh committed:
net: wifi: Add Wi-Fi direct P2P discovery shell command support

Add shell command support for P2P discovery.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

krish2718 commented on modules/hostap/src/supp_api.h:
```diff
@@ -386,4 +386,16 @@ int supplicant_dpp_dispatch(const struct device *dev, struct wifi_dpp_params *pa
+ * @param params P2P parameters
+ * @return 0 for OK; -1 for ERROR
+ */
+int supplicant_p2p(const struct device *dev, struct wifi_p2p_params *params);
```
`supplicant_p2p_oper`?

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -3516,6 +3544,187 @@ static int cmd_wifi_dpp_reconfig(const struct shell *sh, size_t argc, char *argv
+    return 0;
+}
+
+static int cmd_wifi_p2p_find_stop(const struct shell *sh, size_t argc, char *argv[])
```
p2p_stop_find to match wpa_supplicant?

rado17 committed:
drivers: wifi: nrf_wifi: Add RoC support

Add ops for remain-on-channel and cancelling remain-on-channel.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

rado17 committed:
drivers: wifi: nrf_wifi: Allow off channel TX for probe responses

For frames sent down by supplicant in station mode, inform RPU
to allow off-channel transmission. This is needed for sending
P2P probe responses.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

rado17 committed:
drivers: wifi: nrf_wifi: Register frame without match

For frames like Probe Requests, there is no match criterion.
Re-arrange the checks to support registering of frames without
providing any matching info.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

rado17 committed:
modules: hostap: Define heap and stack for P2P support

Increase required heap and stack size for P2P. More stack was required
during WPS negotiation.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

krish2718 commented on drivers/wifi/nrf_wifi/src/fmac_main.c:
```diff
@@ -843,6 +843,8 @@ static int nrf_wifi_drv_main_zep(const struct device *dev)
 	callbk_fns.event_get_wiphy = nrf_wifi_wpa_supp_event_get_wiphy;
 	callbk_fns.mgmt_rx_callbk_fn = nrf_wifi_wpa_supp_event_mgmt_rx_callbk_fn;
 	callbk_fns.get_conn_info_callbk_fn = nrf_wifi_supp_event_proc_get_conn_info;
+	callbk_fns.roc_callbk_fn = nrf_wifi_supp_event_roc_complete;
```
this is not ROC complete, it's ROC ready on channel.

krish2718 commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -1474,6 +1474,35 @@ static int wifi_set_bgscan(uint64_t mgmt_request, struct net_if *iface, void *da
+NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_P2P_OPER, wifi_p2p_oper);
+
+void wifi_mgmt_raise_p2p_device_found_event(struct net_if *iface,
+					     struct wifi_p2p_device_info *device_info)
```
```suggestion
					     struct wifi_p2p_device_info *peer_info)
```
In general, the terminology is, device is local, peer is remote.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -519,6 +520,28 @@ static void handle_wifi_neighbor_rep_complete(struct net_mgmt_event_callback *cb
+		return;
+	}
+
+	if (device_info->device_name[0] != '\0') {
```
```suggestion
	if (!device_info || device_info->device_name[0] == '\0') {
		return;
	}
```

krish2718 reviewed (commented):
* cookie handling
* update wifi-fi mgmt docs (just overview and compile options sections, no need for separate P2P section)
* Update changelog as we have an upcoming release
* Add to twister wherever applicable

krish2718 committed:
modules: hostap: Remove obsolete conditional

We now support a single MbedTLS shim for hostap, so, this extra check is
not needed, we can always use DH5 groups from Mbedtls.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

kapbh committed:
net: wifi: Add Wi-Fi direct P2P connect API support

Add structures and API support for P2P connect.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

krish2718 commented:
the manifest update should be separate commits

kapbh committed:
drivers: nrf_wifi: Add default value to p2p mode Kconfig

The Kconfig NRF70_P2P_MODE should be enabled when
WIFI_NM_WPA_SUPPLICANT_P2P is enabled.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

kapbh committed:
doc: networking: Add Wi-Fi P2P info

Add Wi-Fi P2P mode build command and info.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

rado17 committed:
modules: hostap: Add support for P2P GO mode ops

Enable build time configs required for supporting P2P GO mode.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

rado17 committed:
drivers: wifi: nrf_wifi: Add per-peer authorized flag

Add per-peer authorized parameter. Port authorization command
from supplicant will set this flag and will be used by driver
to allow or nor allow data traffic.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

kapbh committed:
net: wifi: Add Wi-Fi direct P2P connect shell command support

Add shell command support for P2P connect.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

krish2718 commented on drivers/wifi/nrf_wifi/src/net_if.c:
```diff
@@ -436,12 +439,20 @@ int nrf_wifi_if_send(const struct device *dev,
+		ra = nrf_wifi_util_get_ra(sys_dev_ctx->vif_ctx[vif_ctx_zep->vif_idx], nbuf);
+		peer_id = nrf_wifi_fmac_peer_get_id(rpu_ctx_zep->rpu_ctx, ra);
+		if (peer_id == -1) {
+			nrf_wifi_osal_log_err("%s: Invalid peer",
```
log `peer->ra_addr` for debugging. and do you really want a print in the data path? If you can use the Zephyr LOG macros you can use the ratelimited ones.

krish2718 commented on modules/hostap/Kconfig:
```diff
@@ -53,7 +53,7 @@ config HEAP_MEM_POOL_ADD_SIZE_HOSTAP
 	def_int 55000 if WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE && WIFI_CREDENTIALS
 	def_int 48000 if WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE
-	def_int 50000 if WIFI_NM_WPA_SUPPLICANT_P2P
+	def_int 80000 if WIFI_NM_WPA_SUPPLICANT_P2P
```
This is a fixup commit, need to be fixed up.

kapbh committed:
net: wifi: Add API support for P2P GO mode

Add structures and API support for P2P Go mode.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

kapbh committed:
net: wifi: Add Wi-Fi direct P2P GO mode shell command

Add shell commands support for P2P GO mode.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

rado17 committed:
drivers: wifi: nrf_wifi: Add P2P powersave support

Add ops to handle P2P powersave configuration.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

rado17 committed:
drivers: wifi: nrf_wifi: Add cookie handling support

Add cookie event callbacks to track RoC and cancel-RoC
requests and its responses from firmware.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

kapbh committed:
net: wifi: Add API support for P2P power save

Add API support for P2P power save.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

kapbh committed:
net: wifi: Add P2P power save shell command support

Add shell command support for P2P power save.

Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>

rado17 commented on drivers/wifi/nrf_wifi/src/wpa_supp_if.c:
```diff
@@ -1983,6 +1996,141 @@ void nrf_wifi_supp_event_proc_get_conn_info(void *if_priv,
+	roc_info.nrf_wifi_freq_params.channel_type = NRF_WIFI_CHAN_HT20;
+	roc_info.dur = duration;
+
+	status = nrf_wifi_sys_fmac_p2p_roc_start(rpu_ctx_zep->rpu_ctx, vif_ctx_zep->vif_idx, &roc_info);
```
Will wait until a cookie event is received.

rado17 commented on modules/hostap/CMakeLists.txt:
```diff
@@ -394,6 +395,7 @@ zephyr_library_compile_definitions_ifdef(CONFIG_WIFI_NM_WPA_SUPPLICANT_P2P
 zephyr_library_compile_definitions_ifdef(CONFIG_WIFI_NM_WPA_SUPPLICANT_WPS
   CONFIG_WPS
   EAP_WSC
+  EAP_SERVER_WSC
```
Needed by WPS.

rado17 committed:
drivers: wifi: nrf_wifi: Suppress 11b rates in P2P scan

Add an identifier to P2P scan request. RPU can use this to
differentiate it from regular scan requests and suppress
11b rates.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -468,6 +477,38 @@ enum net_event_wifi_cmd {
+	/** Manufacturer (max 64 chars + null terminator) */
+	char manufacturer[65];
+	/** Model name (max 32 chars + null terminator) */
+	char model_name[33];
```
and for these too

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -2586,3 +2595,235 @@ int supplicant_config_params(const struct device *dev, struct wifi_config_params
+	} else if (strncmp(line, "level=", 6) == 0) {
+		pos = line + 6;
+		/* Skip extraction for numeric values, atoi handles it */
+		info->rssi = (int8_t)atoi(pos);
```
We are mixing `atoi()` and `strtol()` and the atoi one cannot detect errors in the input string. I suggest we use only `strtol()` and make sure that the string is a valid one (now we ignore the errors).

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -2586,3 +2595,235 @@ int supplicant_config_params(const struct device *dev, struct wifi_config_params
+		break;
+
+	case WIFI_P2P_PEER:
+		char addr[P2P_ADDR_SIZE];
```
Could you add a block here i.e., `{` because you are defining variables.

jukkar commented on modules/hostap/src/supp_api.c:
```diff
@@ -2586,3 +2595,235 @@ int supplicant_config_params(const struct device *dev, struct wifi_config_params
+				return -EINVAL;
+			}
+
+			snprintf(addr_str, sizeof(addr_str), "%02x:%02x:%02x:%02x:%02x:%02x",
```
please replace all `snprintf()` by `snprintk()` (it uses less resources)

D-Triveni committed:
drivers: wifi: nrf_wifi: Fix invalid pointer access

Fix pointer dereferencing by accessing the pointer only
after initialization, preventing fault exceptions.

Signed-off-by: Triveni Danda <triveni.danda@nordicsemi.no>

krish2718 committed:
modules: hostap: Fix SoF in softap mode

Due to recent changes, the stack usage is increased, fix the SoF.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
drivers: nrf_wifi: Workaround for failing tests

Test automation framework fails if there are any error prints, and we
had seen issues with networking that sends packets to the driver before
the assocation is up (either not checking dormant status or in a window
where dormant status is being updated).

Add a workaround to suppress the print till the issue root cause is
fixed.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

jukkar commented on modules/hostap/Kconfig:
```diff
@@ -61,6 +62,7 @@ endif # WIFI_NM_WPA_SUPPLICANT_GLOBAL_HEAP
 
 config WIFI_NM_WPA_SUPPLICANT_THREAD_STACK_SIZE
 	int "Stack size for wpa_supplicant thread"
+	default 10000 if WIFI_NM_WPA_SUPPLICANT_P2P
```
Yeah, this is extremely high. Can this be adjusted?

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -1522,6 +1528,50 @@ struct wifi_p2p_params {
+	/** Group remove specific parameters */
+	struct {
+		/** Interface name (e.g., "wlan0") */
+		char ifname[CONFIG_NET_INTERFACE_NAME_LEN + 1];
```
We should use `NET_IFNAMSIZ` instead of `CONFIG_NET_INTERFACE_NAME_LEN` as the config option might be missing. I suggest we have a separate PR that changes all the use of `CONFIG_NET_INTERFACE_NAME_LEN` in `modules/hostap` as there are many found there. @rado17 or @krish2718 are you able to send a PR for this after this one is merged?

krish2718 committed:
drivers: nrf_wifi: Fix the port authorization logic

For non-STA mode the vif level authorized flag is unused, simply allow
all group traffic and for unicast traffic check respective peer status.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
drivers: nrf_wifi: Set VIF type for non-STA

This is used in SoftAP/P2P GO modes.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

jukkar commented on drivers/wifi/nrf_wifi/src/fmac_main.c:
```diff
@@ -101,6 +101,61 @@ static const unsigned int rx3_buf_sz = CONFIG_NRF70_RX_MAX_DATA_SIZE;
+ *
+ * @return Pointer to the buffer on success, NULL on failure.
+ */
+char *nrf_wifi_sprint_ll_addr_buf(const uint8_t *ll, uint8_t ll_len,
```
If we are printing MAC address, why do we need to check 8 or 2 bytes long addresses?

rado17 replied:
Done

jukkar commented on drivers/wifi/nrf_wifi/src/wpa_supp_if.c:
```diff
@@ -1698,7 +1717,7 @@ int nrf_wifi_supp_register_frame(void *if_priv,
 	struct nrf_wifi_umac_mgmt_frame_info frame_info;
 
-	if (!if_priv || !match || !match_len) {
+	if (!if_priv) {
```
Please make checks explicit, so `if_priv == NULL` here and all over the place

rado17 replied:
Resolved now

rado17 committed:
drivers: wifi: nrf_wifi: Use filter_ssid for connect scan

Modify the scan logic to use filter_ssids for connect scan
and use ssids for other cases.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>
